### DISCUSSION TITLE
`new Worker()` should automatically gain `worker` target environment (vis-a-vis package.json exports)

### DISCUSSION THREAD

**Discussion by nemosmithasf** [webpack/webpack#16900] [Q&A]
## Feature request

**What is the expected behavior?**
When running `new Worker()`, all the imports in that entry-point should use `worker` [target environment](https://webpack.js.org/guides/package-exports/#target-environment)

Instead, it seems the default is just browser.

**What is motivation or use case for adding/changing the behavior?**
Makes it easier to build libraries that are compatible with workers (i.e. same motivation as browser/node conditional exports)

**How should this be implemented in your opinion?**
As stated in title.

**Are you willing to work on this yourself?**
No

**Comment by alexander-akait** [comment]
Please clarify what do you mean? We already respect `worker` in `exports`, will be great if you provide reproducible example

**Reply by nemosmithasf** [reply]
Consider the `snippet`:

```
import conditional_entry from "libraryA"
```

And `libraryA` with package.json of:

```
{
  exports:{
    worker: "worker_entry.js",
    browser: "browser_entry.js"
  }
}
```

if I run `snippet` in a browser environment, it would import `browser_entry.js`. If I run `snippet` in a Worker, it would be ideal if we can get `worker_entry.js`- because right now I believe it still selects `browser_entry.js`.

I could be wrong, so please correct me on this. But from what I gathered, the only way to get `worker_entry.js` is to explicitly set the `target` to `worker`... but since we know when an entry-point is being used in a worker (e.g. `new Worker(new URL(...))`), shouldn't we automatically set the `target` to `worker`?

**Reply by alexander-akait** [reply]
It always should work as you desribed, can you create reproducible test repo with the problem?

**Reply by jarrodldavis** [reply]
@alexander-akait I am having the same issue and I have created a reproduction of the problem here: https://github.com/jarrodldavis/webpack-worker-repro

**Reply by jarrodldavis** [reply]
It looks like the issue stems from `WorkerPlugin` creating an async/codesplit chunk:

https://github.com/webpack/webpack/blob/27cf3e59f5f289dfc4d76b7a1df2edbc4e651589/lib/dependencies/WorkerPlugin.js#L336-L354

This will just use the same `target` as the rest of the compilation, unfortunately, since there's no way to override `target` for a specific part of the module graph.

I've also tried setting `resolve.byDependency.worker.conditionNames` in `webpack.config.js`, but that only works for resolving the initial module of the worker, rather than the entire subgraph that is used by the worker.

If I am not mistaken, I think the only way to fix this is to have `WorkerPlugin` create a child compiler to compile the worker (sub)graph with the `target` option set to `webworker`.
